submit-api/src/SubmitAPI/Internal/Transaction.hs

```diff
@@ -0,0 +1,190 @@
+module SubmitAPI.Internal.Transaction where
+
+import           RIO
+import qualified Data.Text                 as T
+import qualified Data.Map.Strict           as Map
+import           Codec.Serialise
+import           Data.ByteString.Lazy      (toStrict)
+import           Data.Text.Prettyprint.Doc (Pretty(..))
+import qualified Data.Set                  as Set
+
+import           Cardano.Api                    hiding (TxBodyError)
+import           Cardano.Api.Shelley            (ProtocolParameters)
+import           Cardano.Slotting.Time          (SystemStart)
+import qualified Ledger                         as P
+import qualified Plutus.Contract.CardanoAPI     as Interop
+import           Plutus.Contract.CardanoAPITemp as Interop
+import qualified Ledger.Ada                     as Ada
+
+import qualified CardanoTx.Models   as Sdk
+import           CardanoTx.ToPlutus
+import           NetworkAPI.Env
+
+signTx
+  :: TxBody AlonzoEra
+  -> [ShelleyWitnessSigningKey]
+  -> Tx AlonzoEra
+signTx body keys =
+  makeSignedTransaction wits body
+    where wits = keys <&> makeShelleyKeyWitness body
+
+buildBalancedTx
+  :: MonadThrow f
+  => SystemEnv
+  -> Sdk.ChangeAddress
+  -> Set.Set Sdk.FullCollateralTxIn
+  -> Sdk.TxCandidate
+  -> f (BalancedTxBody AlonzoEra)
+buildBalancedTx SystemEnv{..} defaultChangeAddr collateral txc@(Sdk.TxCandidate{..}) = do
+  let era          = AlonzoEra
+      eraInMode    = AlonzoEraInCardanoMode
+      witOverrides = Nothing
+
+  txBody     <- buildTxBodyContent pparams network collateral txc
+  inputsMap  <- mkInputsUTxO network (Set.elems txCandidateInputs)
+  changeAddr <- absorbError $ case txCandidateChangePolicy of
+    Just (Sdk.ReturnTo addr) -> Interop.toCardanoAddress network addr
+    _                        -> Interop.toCardanoAddress network $ Sdk.getAddress defaultChangeAddr
+
+  absorbBalancingError $ makeTransactionBodyAutoBalance eraInMode sysstart eraHistory pparams pools inputsMap txBody changeAddr witOverrides
+    where
+      absorbBalancingError (Left e)  = throwM $ BalancingError $ T.pack $ show e
+      absorbBalancingError (Right a) = pure a
+
+buildTxBodyContent
+  :: MonadThrow f
+  => ProtocolParameters
+  -> NetworkId
+  -> Set.Set Sdk.FullCollateralTxIn
+  -> Sdk.TxCandidate
+  -> f (TxBodyContent BuildTx AlonzoEra)
+buildTxBodyContent protocolParams network collateral Sdk.TxCandidate{..} = do
+  txIns           <- buildTxIns $ Set.elems txCandidateInputs
+  txInsCollateral <- buildTxCollateral $ Set.elems collateral
+  txOuts          <- buildTxOuts network txCandidateOutputs
+  txFee           <- absorbError $ Interop.toCardanoFee dummyFee
+  txValidityRange <- absorbError $ Interop.toCardanoValidityRange txCandidateValidRange
+  txMintValue     <- absorbError $ Interop.toCardanoMintValue (Sdk.unMintValue txCandidateValueMint) txCandidateMintPolicies
+  txData          <- collectInputsData $ Set.elems txCandidateInputs
+
+  pure $ TxBodyContent
```

Comment:
Maybe create smart constructors?

Reply:
For external model? Why?

Reply:
cause there are lots of unused fields with default values

Reply:
But that's the only place where the construction of TxBodyContent happens. What's the point in moving this to a separate method and then use here?